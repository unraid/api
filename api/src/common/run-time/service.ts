/* eslint-disable new-cap */
import { Number, String, Boolean, Record, Null, type Static } from 'runtypes';

const uptime = Record({
	timestamp: String.Or(Null),
});

export const Service = Record({
	name: String.Or(Null),
	online: Boolean.Or(Null),
	uptime: Number.Or(uptime).Or(Null),
	version: String.Or(Null),
});

// eslint-disable-next-line @typescript-eslint/no-redeclare
export type Service = Static<typeof Service>;
