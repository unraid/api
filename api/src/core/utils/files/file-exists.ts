import { access } from 'fs/promises';
import { accessSync } from 'fs';
import { F_OK } from 'node:constants';

export const fileExists = async (path: string) => access(path, F_OK).then(() => true).catch(() => false);
export const fileExistsSync = (path: string) => {
	try {
		accessSync(path, F_OK);
		return true;
	} catch (error: unknown) {
		return false;
	}
};
