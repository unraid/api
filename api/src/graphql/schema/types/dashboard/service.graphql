
type DashboardServiceUptime {
  timestamp: DateTime
}

type DashboardService {
  name: String
  online: Boolean
  uptime: DashboardServiceUptime
  version: String
}

input DashboardServiceUptimeInput {
  timestamp: DateTime!
}

input DashboardServiceInput {
  name: String!
  online: Boolean!
  uptime: DashboardServiceUptimeInput
  version: String!
}