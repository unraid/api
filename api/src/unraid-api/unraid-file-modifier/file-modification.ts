import { Logger } from '@nestjs/common';
import { constants } from 'fs';
import { access, readFile, unlink, writeFile } from 'fs/promises';
import { basename, dirname, join } from 'path';

import { applyPatch, parsePatch, reversePatch } from 'diff';

export interface ShouldApplyWithReason {
    shouldApply: boolean;
    reason: string;
}

// Convert interface to abstract class with default implementations
export abstract class FileModification {
    abstract id: string;
    public abstract readonly filePath: string;

    protected constructor(protected readonly logger: Logger) {}

    // This is the main method that child classes need to implement
    protected abstract generatePatch(): Promise<string>;

    private getPatchFilePath(targetFile: string): string {
        const dir = dirname(targetFile);
        const filename = `${basename(targetFile)}.patch`;
        return join(dir, filename);
    }

    private async savePatch(patchResult: string): Promise<void> {
        const patchFile = this.getPatchFilePath(this.filePath);
        await writeFile(patchFile, patchResult, 'utf8');
    }

    private async loadSavedPatch(targetFile: string): Promise<string | null> {
        const patchFile = this.getPatchFilePath(targetFile);
        try {
            await access(patchFile, constants.R_OK);
            return await readFile(patchFile, 'utf8');
        } catch {
            return null;
        }
    }

    private async getPregeneratedPatch(): Promise<string | null> {
        const patchResults = await import.meta.glob('./modifications/patches/*.patch', {
            query: '?raw',
            import: 'default',
        });

        if (patchResults[`./modifications/patches/${this.id}.patch`]) {
            const loader = Object.values(patchResults)[0];
            const fileContents = await loader();
            this.logger.debug(`Loaded pregenerated patch for ${this.id}`);
            if (typeof fileContents !== 'string') {
                this.logger.error('Invalid patch format on patch: ' + this.id);
                return null;
            }
            return fileContents;
        }

        return null;
    }

    private async applyPatch(patchContents: string): Promise<void> {
        const currentContent = await readFile(this.filePath, 'utf8');
        const parsedPatch = parsePatch(patchContents)[0];
        const results = applyPatch(currentContent, parsedPatch);
        if (results === false) {
            throw new Error(`Failed to apply patch to ${this.filePath}`);
        }
        await writeFile(this.filePath, results);
    }

    // Default implementation of apply that uses the patch
    async apply(): Promise<void> {
        // First attempt to apply the patch that was generated
        const staticPatch = await this.getPregeneratedPatch();
        if (staticPatch) {
            try {
                await this.applyPatch(staticPatch);
                await this.savePatch(staticPatch);
                return;
            } catch (error) {
                this.logger.error(
                    `Failed to apply static patch to ${this.filePath}, continuing with dynamic patch`
                );
            }
        }
        const patchContents = await this.generatePatch();
        await this.applyPatch(patchContents);
        await this.savePatch(patchContents);
    }

    // Update rollback to use the shared utility
    async rollback(): Promise<void> {
        let patch: string;

        // Try to load saved patch first
        const savedPatch = await this.loadSavedPatch(this.filePath);
        if (savedPatch) {
            this.logger.debug(`Using saved patch file for ${this.id}`);
            patch = savedPatch;
        } else {
            this.logger.debug(`No saved patch found for ${this.id}, generating new patch`);
            const patchContents = await this.generatePatch();
            patch = patchContents;
        }

        const currentContent = await readFile(this.filePath, 'utf8');
        const parsedPatch = parsePatch(patch)[0];

        if (!parsedPatch || !parsedPatch.hunks || parsedPatch.hunks.length === 0) {
            throw new Error('Invalid or empty patch content');
        }

        const reversedPatch = reversePatch(parsedPatch);
        const results = applyPatch(currentContent, reversedPatch);

        if (results === false) {
            throw new Error(`Failed to rollback patch from ${this.filePath}`);
        }

        await writeFile(this.filePath, results);

        // Clean up the patch file after successful rollback
        try {
            const patchFile = this.getPatchFilePath(this.filePath);
            await access(patchFile, constants.W_OK);
            await unlink(patchFile);
        } catch {
            // Ignore errors when trying to delete the patch file
        }
    }

    // Default implementation that can be overridden if needed
    async shouldApply(): Promise<ShouldApplyWithReason> {
        return {
            shouldApply: true,
            reason: 'Default behavior is to always apply modifications',
        };
    }
}
