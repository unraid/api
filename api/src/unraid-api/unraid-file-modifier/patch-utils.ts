import { readFile, writeFile } from 'fs/promises';
import { applyPatch, parsePatch, reversePatch } from 'diff';

export async function rollbackPatch(targetFile: string, patch: string): Promise<void> {
    const currentContent = await readFile(targetFile, 'utf8');
    const parsedPatch = parsePatch(patch)[0];
    
    if (!parsedPatch || !parsedPatch.hunks || parsedPatch.hunks.length === 0) {
        throw new Error('Invalid or empty patch content');
    }

    const reversedPatch = reversePatch(parsedPatch);
    const results = applyPatch(currentContent, reversedPatch);

    if (results === false) {
        throw new Error(`Failed to rollback patch from ${targetFile}`);
    }

    await writeFile(targetFile, results);
} 