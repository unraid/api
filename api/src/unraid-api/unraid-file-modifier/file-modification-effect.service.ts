import { Injectable } from '@nestjs/common';

import { NginxService } from '@app/unraid-api/nginx/nginx.service.js';
import { ModificationEffect } from '@app/unraid-api/unraid-file-modifier/file-modification.js';

@Injectable()
export class FileModificationEffectService {
    constructor(private readonly nginxService: NginxService) {}
    async runEffect(effect: ModificationEffect): Promise<void> {
        switch (effect) {
            case 'nginx:reload':
                await this.nginxService.reload();
                break;
        }
    }
}
