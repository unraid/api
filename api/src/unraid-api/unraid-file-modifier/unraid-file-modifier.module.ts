import { Module } from '@nestjs/common';
import { FileModificationService } from './unraid-file-modifier.service';

@Module({
  providers: [FileModificationService]
})
export class UnraidFileModifierModule {}
