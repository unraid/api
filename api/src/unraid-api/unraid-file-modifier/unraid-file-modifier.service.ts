import { Injectable, Logger, OnModuleDestroy, OnModuleInit } from '@nestjs/common';
import AuthRequestModification from '@app/unraid-api/unraid-file-modifier/modifications/auth-request.modification';
import DefaultPageLayoutModification from '@app/unraid-api/unraid-file-modifier/modifications/default-page-layout.modification';
import { LogRotateModification } from '@app/unraid-api/unraid-file-modifier/modifications/log-rotate.modification';
import NotificationsPageModification from '@app/unraid-api/unraid-file-modifier/modifications/notifications-page.modification';
import SSOFileModification from '@app/unraid-api/unraid-file-modifier/modifications/sso.modification';
import { FileModification } from '@app/unraid-api/unraid-file-modifier/file-modification';

@Injectable()
export class UnraidFileModificationService implements OnModuleInit, OnModuleDestroy {
    private readonly logger = new Logger(UnraidFileModificationService.name);
    private appliedModifications: FileModification[] = [];

    async onModuleInit() {
        try {
            this.logger.log('Loading file modifications...');
            const mods = await this.loadModifications();
            await this.applyModifications(mods);
        } catch (err) {
            if (err instanceof Error) {
                this.logger.error(`Failed to apply modifications: ${err.message}`);
            } else {
                this.logger.error('Failed to apply modifications: Unknown error');
            }
        }
    }

    async onModuleDestroy() {
        this.logger.log('Rolling back all modifications...');
        await this.rollbackAll();
    }

    async loadModifications(): Promise<FileModification[]> {
        const modifications: FileModification[] = [];
        const modificationClasses: Array<new (logger: Logger) => FileModification> = [
            LogRotateModification,
            AuthRequestModification,
            SSOFileModification,
            DefaultPageLayoutModification,
            NotificationsPageModification,
        ];
        for (const ModificationClass of modificationClasses) {
            const instance = new ModificationClass(this.logger);
            modifications.push(instance);
        }
        return modifications;
    }

    async applyModifications(modifications: FileModification[]): Promise<void> {
        for (const modification of modifications) {
            await this.applyModification(modification);
        }
    }

    async applyModification(modification: FileModification): Promise<void> {
        try {
            const shouldApplyWithReason = await modification.shouldApply();
            if (shouldApplyWithReason.shouldApply) {
                this.logger.log(
                    `Applying modification: ${modification.id} - ${shouldApplyWithReason.reason}`
                );
                await modification.apply();
                this.appliedModifications.push(modification);
                this.logger.log(`Modification applied successfully: ${modification.id}`);
            } else {
                this.logger.log(
                    `Skipping modification: ${modification.id} - ${shouldApplyWithReason.reason}`
                );
            }
        } catch (error) {
            if (error instanceof Error) {
                this.logger.error(
                    `Failed to apply modification: ${modification.id}: ${error.message}`,
                    error.stack
                );
            } else {
                this.logger.error(`Failed to apply modification: ${modification.id}: Unknown error`);
            }
        }
    }

    async rollbackAll(): Promise<void> {
        // Process modifications in reverse order
        for (const modification of [...this.appliedModifications].reverse()) {
            try {
                this.logger.log(`Rolling back modification: ${modification.id}`);
                await modification.rollback();
                this.logger.log(`Successfully rolled back modification: ${modification.id}`);
            } catch (error) {
                if (error instanceof Error) {
                    this.logger.error(`Failed to roll back modification: ${error.message}`);
                } else {
                    this.logger.error('Failed to roll back modification: Unknown error');
                }
            }
        }
        this.appliedModifications = [];
    }
}
