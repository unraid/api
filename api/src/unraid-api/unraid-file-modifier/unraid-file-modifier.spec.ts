import { Logger } from '@nestjs/common';
import { Test, TestingModule } from '@nestjs/testing';
import { promises as fs } from 'fs';
import { join } from 'path';

import { createPatch } from 'diff';
import { afterEach, beforeEach, describe, expect, it, vi } from 'vitest';

import { fileExistsSync } from '@app/core/utils/files/file-exists';
import {
    FileModification,
    ShouldApplyWithReason,
} from '@app/unraid-api/unraid-file-modifier/file-modification';
import { UnraidFileModificationService } from '@app/unraid-api/unraid-file-modifier/unraid-file-modifier.service';

const FIXTURE_PATH = join(__dirname, 'modifications', '__test__', '__fixtures__', 'text-patch-file.txt');
const ORIGINAL_CONTENT = 'original';

class TestFileModification extends FileModification {
    id = 'test';
    public readonly filePath: string = FIXTURE_PATH;

    protected async generatePatch(overridePath?: string): Promise<string> {
        return createPatch(overridePath ?? 'text-patch-file.txt', ORIGINAL_CONTENT, 'modified');
    }

    async shouldApply(): Promise<ShouldApplyWithReason> {
        return { shouldApply: true, reason: 'Always Apply this mod' };
    }
}

describe.sequential('FileModificationService', () => {
    let mockLogger: {
        log: ReturnType<typeof vi.fn>;
        error: ReturnType<typeof vi.fn>;
        warn: ReturnType<typeof vi.fn>;
        debug: ReturnType<typeof vi.fn>;
        verbose: ReturnType<typeof vi.fn>;
    };
    let service: UnraidFileModificationService;
    let logger: Logger;

    beforeEach(async () => {
        // Create/reset the fixture file before each test
        await fs.writeFile(FIXTURE_PATH, ORIGINAL_CONTENT);

        mockLogger = {
            log: vi.fn(),
            error: vi.fn(),
            warn: vi.fn(),
            debug: vi.fn(),
            verbose: vi.fn(),
        };

        vi.spyOn(Logger.prototype, 'log').mockImplementation(mockLogger.log);
        vi.spyOn(Logger.prototype, 'error').mockImplementation(mockLogger.error);
        vi.spyOn(Logger.prototype, 'warn').mockImplementation(mockLogger.warn);
        vi.spyOn(Logger.prototype, 'debug').mockImplementation(mockLogger.debug);
        vi.spyOn(Logger.prototype, 'verbose').mockImplementation(mockLogger.verbose);

        logger = new Logger('test');

        const module: TestingModule = await Test.createTestingModule({
            providers: [UnraidFileModificationService],
        }).compile();

        service = module.get<UnraidFileModificationService>(UnraidFileModificationService);
    });

    it('should be defined', () => {
        expect(service).toBeDefined();
    });

    it('should load modifications', async () => {
        const mods = await service.loadModifications();
        expect(mods.length).toBeGreaterThan(0);
    });

    it('should apply modifications', async () => {
        const mod = new TestFileModification(logger);
        await expect(service.applyModification(mod)).resolves.toBe(undefined);
    });

    it('should apply modification if file does not exist', async () => {
        const mod = new TestFileModification(logger);
        // @ts-expect-error - This is a protected method, but we need to mock it
        mod.generatePatch = vi.fn().mockResolvedValue(createPatch(FIXTURE_PATH, '', 'modified'));
        await fs.unlink(FIXTURE_PATH);
        await expect(service.applyModification(mod)).resolves.toBe(undefined);
        expect(mockLogger.warn).toHaveBeenCalledWith('Could not load pregenerated patch for: test');
        expect(mockLogger.log).toHaveBeenCalledWith(
            'Applying modification: test - Always Apply this mod'
        );
        expect(mockLogger.log).toHaveBeenCalledWith('Modification applied successfully: test');
        const content = await fs.readFile(FIXTURE_PATH, 'utf-8');
        expect(content).toBe('modified');
        await service.rollbackAll();
        expect(fileExistsSync(FIXTURE_PATH)).toBe(false);
        expect(mockLogger.log).toHaveBeenCalledWith('Rolling back modification: test');
        expect(mockLogger.log).toHaveBeenCalledWith('Successfully rolled back modification: test');
    });

    it('should not rollback any mods without loaded', async () => {
        await expect(service.rollbackAll()).resolves.toBe(undefined);
    });

    it('should rollback all mods', async () => {
        await service.loadModifications();
        const initialContent = await fs.readFile(FIXTURE_PATH, 'utf-8');
        expect(initialContent).toBe(ORIGINAL_CONTENT);

        const mod = new TestFileModification(logger);

        await service.applyModification(mod);
        const modifiedContent = await fs.readFile(FIXTURE_PATH, 'utf-8');
        expect(modifiedContent).toBe('modified');

        await service.rollbackAll();
        const rolledBackContent = await fs.readFile(FIXTURE_PATH, 'utf-8');
        expect(rolledBackContent).toBe(ORIGINAL_CONTENT);

        expect(mockLogger.warn).toHaveBeenCalledWith('Could not load pregenerated patch for: test');
        expect(mockLogger.log.mock.calls).toEqual([
            ['RootTestModule dependencies initialized'],
            ['Applying modification: test - Always Apply this mod'],
            ['Modification applied successfully: test'],
            ['Rolling back modification: test'],
            ['Successfully rolled back modification: test'],
        ]);
    });

    it('should handle errors during dual application', async () => {
        await service.loadModifications();
        const initialContent = await fs.readFile(FIXTURE_PATH, 'utf-8');
        expect(initialContent).toBe(ORIGINAL_CONTENT);

        const mod = new TestFileModification(logger);

        await service.applyModification(mod);

        expect(mockLogger.log.mock.calls).toEqual([
            ['RootTestModule dependencies initialized'],
            ['Applying modification: test - Always Apply this mod'],
            ['Modification applied successfully: test'],
        ]);

        // Now apply again and ensure the contents don't change
        await service.applyModification(mod);
        const errorMessage = mockLogger.warn.mock.calls[0][0];
        expect(errorMessage).toContain('Could not load pregenerated patch for: test');
    });

    afterEach(async () => {
        await service.rollbackAll();
        vi.clearAllMocks();
    });
});
