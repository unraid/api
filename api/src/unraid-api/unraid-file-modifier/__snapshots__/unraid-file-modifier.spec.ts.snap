// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FileModificationService > should apply modifications 1`] = `[Error: Application not implemented.]`;
